src/common/RWRef.h

```diff
@@ -0,0 +1,244 @@
+ * the "require" state is matched, or set a flag to tell the callers
+ * that the "require" state is not matched and also there is no any
+ * wait mechanism for "readers" to wait the state until it matches. It
+ * will let the callers determine what to do next.
+ *
+ * The usage, such as in libcephfs's client/Client.cc case:
+ *
+ * The Readers:
+ *
+ *   For the ll_read()/ll_write(), etc fucntions, they will work as
+ *   "readers", in the beginning they just need to define a RWRef
+ *   object and in RWRef constructor it will check if the state is
+ *   MOUNTED or MOUTING, if not it will fail and return directly with
+ *   doing nothing, or it will increase the reference and continue.
+ *   And when destructing the RWRef object, in the RWRef destructor
+ *   it will decrease the reference and notify the "writers" who maybe
+ *   waiting.
+ *
+ * The Writers:
+ *
+ *   And for the _unmount() function , as a "writer", in the beginning
+ *   it will also just need to define a RWRef object and in RWRef
+ *   constructor it will update the state to next stage first, which then
+ *   will fail all the new comming "readers", and then wait for all the
+ *   "readers" to finish.
+ *
+ * With this we can get rid of the locks for all the "readers" and they
+ * can run in parallel. And we won't have any potential deadlock issue
+ * with RWRef, such as:
+ *
+ * With RWLock:
+ *
+ *     ThreadA:                           ThreadB:
+ *
+ *     write_lock<RWLock1>.lock();        another_lock.lock();
+ *     state = NEXT_STATE;                ...
+ *     another_lock.lock();               read_lock<RWLock1>.lock();
+ *     ...                                if (state == STATE) {
+ *                                          ...
+ *                                        }
+ *                                        ...
+ *
+ * With RWRef:
+ *
+ *     ThreadA:                           ThreadB:
+ *
+ *     w = RWRef(myS, NEXT_STATE, false); another_lock.lock();
+ *     another_lock.lock();               r = RWRef(myS, STATE);
+ *     ...                                if (r.is_state_satisfied()) {
+ *                                          ...
+ *                                        }
+ *                                        ...
+ *
+ * And also in ThreadA, if it needs to do the cond.wait(&another_lock),
+ * it will goto sleep by holding the write_lock<RWLock1> for the RWLock
+ * case, if the ThreadBs are for some IOs, they may stuck for a very long
+ * time that may get timedout in the uplayer which may keep retrying.
+ * With the RWRef, the ThreadB will fail or continue directly without any
+ * stuck, and the uplayer will knew what next to do quickly.
+ */
+
+#ifndef CEPH_RWRef_Posix__H
+#define CEPH_RWRef_Posix__H
+
+#include <string>
+#include "include/ceph_assert.h"
+#include "common/ceph_mutex.h"
+
+/* The status mechanism info */
+struct RWRefState {
+  private:
+    ceph::mutex lock;
+    ceph::condition_variable cond;
+    uint64_t reader_cnt = 0;
```

Comment:
private declarations should go at the end, see: https://google.github.io/styleguide/cppguide.html#Declaration_Order